rclcpp/include/rclcpp/loaned_message.hpp

```diff
@@ -0,0 +1,170 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__LOANED_MESSAGE_HPP_
+#define RCLCPP__LOANED_MESSAGE_HPP_
+
+#include <memory>
+#include <utility>
+
+#include "rclcpp/logging.hpp"
+#include "rclcpp/publisher_base.hpp"
+
+#include "rcl/allocator.h"
+#include "rcl/publisher.h"
+
+namespace rclcpp
+{
+
+template<typename MessageT, typename AllocatorT = std::allocator<void>>
+class LoanedMessage
+{
+  using MessageAllocatorTraits = allocator::AllocRebind<MessageT, AllocatorT>;
+  using MessageAllocator = typename MessageAllocatorTraits::allocator_type;
+
+protected:
+  const rclcpp::PublisherBase * pub_;
+
+  std::unique_ptr<MessageT> message_;
```

Comment:
we discussed this offline that with having a `std::unique_ptr` the class becomes implicitly non-trivial copyable. Do having a move constructor is really only needed for RVO.

Reply:
Ok, that's fine, it just seems a bit odd to be using `reset` and `release` in the constructor/destructor of this class rather than using the deleter of the unique pointer.

You could also just disable the copy constructors and use a raw pointer.